docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,524 @@
+
+In this case, when developers are confronted with the issues described above they can require dependencies to update in
+sync (what we do now) or attempt some ad-hoc potentially hacky solution.
+
+One approach is to ditch go semantic import versioning (SIV) altogether. Some people have commented that go's SIV
+(i.e. changing the import path to `foo/v2`, `foo/v3`, etc.) is too restrictive and that it should be optional. The
+golang maintainers disagree and only officially support semantic import versioning. We could, however, take the
+contrarian perspective and get more flexibility by using 0.x-based versioning basically forever. 
+
+Module version compatibility could then be achieved using go.mod replace directives to pin dependencies to specific
+compatible 0.x versions. For instance if we knew `foo` 0.2 and 0.3 were both compatible with `bar` 0.3 and 0.4, we
+could use replace directives in our go.mod to stick to the versions of `foo` and `bar` we want. This would work as
+long as the authors of `foo` and `bar` avoid incompatible breaking changes between these modules.
+
+Or, if developers choose to use semantic import versioning, they can attempt the naive solution described above
+and would also need to use special tags and replace directives to make sure that modules are pinned to the correct
+versions.
+
+Note, however, that all of these ad-hoc approaches, would be vulnerable to the minor version compatibility issues
+described above unless [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering)
+is properly addressed.
+
+### Approach D) Avoid protobuf generated code in public APIs
+
+An alternative approach would be to avoid protobuf generated code in public module APIs. This would help avoid the
+discrepancy between state machine versions and client API versions at the module to module boundaries. It would mean
+that we wouldn't do inter-module message passing based on ADR 033, but rather stick to the existing keeper approach
+and take it one step further by avoiding any protobuf generated code in the keeper interface methods.
+
+Using this approach, our `foo.Keeper.DoSomething` method wouldn't have the generated `MsgDoSomething` struct (which
+comes from the protobuf API), but instead positional parameters. Then in order for `foo/v2` to support the `foo/v1`
+keeper it would simply need to implement both the v1 and v2 keeper APIs. The `DoSomething` method in v2 could have the
+additional `condition` parameter, but this wouldn't be present in v1 at all so there would be no danger of a client
+accidentally setting this when it isn't available. 
+
+So this approach would avoid the challenge around minor version incompatibilities because the existing module keeper
+API would not get new fields when they are added to protobuf files.
+
+Taking this approach, however, would likely require making all protobuf generated code internal in order to prevent
```

Comment:
We need to access proto API. Mostly in the client code. For tools written in Go, I think it's better to import a generate code, rather then ask all tools to manage that by themselves.
Moreover, for cross-chain compatibility, such tools will need to support multiple versions. Hence we need stable versions (v1, v2, ....) rather than v0.1, v0.2 ... (in Go we can import module/v1 and module/v2 at the same time, but can't do it with v0.1 and v0.2).